File: mobile/lib/models/leaderboard.dart
class LeaderboardEntry {
  final int rank;
  final String userId;
  final String username;
  final String? displayName;
  final String? avatarUrl;
  final int points;
  final double? masteryPercentage;

  LeaderboardEntry({
    required this.rank,
    required this.userId,
    required this.username,
    this.displayName,
    this.avatarUrl,
    required this.points,
    this.masteryPercentage,
  });

  factory LeaderboardEntry.fromJson(Map<String, dynamic> json) {
    return LeaderboardEntry(
      rank: json['rank'] ?? 0,
      userId: json['user_id'] ?? '',
      username: json['username'] ?? '',
      displayName: json['display_name'],
      avatarUrl: json['avatar_url'],
      points: json['points'] ?? 0,
      masteryPercentage: json['mastery_percentage']?.toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'rank': rank,
      'user_id': userId,
      'username': username,
      'display_name': displayName,
      'avatar_url': avatarUrl,
      'points': points,
      'mastery_percentage': masteryPercentage,
    };
  }

  String get displayNameOrUsername => displayName ?? username;
}

class LeaderboardResponse {
  final String scope;
  final String? categoryId;
  final List<LeaderboardEntry> entries;
  final int? userRank;
  final int totalUsers;

  LeaderboardResponse({
    required this.scope,
    this.categoryId,
    required this.entries,
    this.userRank,
    required this.totalUsers,
  });

  factory LeaderboardResponse.fromJson(Map<String, dynamic> json) {
    return LeaderboardResponse(
      scope: json['scope'] ?? 'weekly',
      categoryId: json['category_id'],
      entries: (json['entries'] as List<dynamic>?)
              ?.map((e) => LeaderboardEntry.fromJson(e))
              .toList() ??
          [],
      userRank: json['user_rank'],
      totalUsers: json['total_users'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'scope': scope,
      'category_id': categoryId,
      'entries': entries.map((e) => e.toJson()).toList(),
      'user_rank': userRank,
      'total_users': totalUsers,
    };
  }
}

class AppNotification {
  final String id;
  final String userId;
  final String title;
  final String body;
  final String notificationType;
  final String? actionUrl;
  final bool isRead;
  final bool isPushed;
  final DateTime createdAt;
  final DateTime? expiresAt;
  final DateTime? readAt;

  AppNotification({
    required this.id,
    required this.userId,
    required this.title,
    required this.body,
    required this.notificationType,
    this.actionUrl,
    required this.isRead,
    required this.isPushed,
    required this.createdAt,
    this.expiresAt,
    this.readAt,
  });

  factory AppNotification.fromJson(Map<String, dynamic> json) {
    return AppNotification(
      id: json['id'],
      userId: json['user_id'],
      title: json['title'] ?? '',
      body: json['body'] ?? '',
      notificationType: json['notification_type'] ?? 'general',
      actionUrl: json['action_url'],
      isRead: json['is_read'] ?? false,
      isPushed: json['is_pushed'] ?? false,
      createdAt: DateTime.parse(json['created_at']),
      expiresAt: json['expires_at'] != null
          ? DateTime.parse(json['expires_at'])
          : null,
      readAt: json['read_at'] != null ? DateTime.parse(json['read_at']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'title': title,
      'body': body,
      'notification_type': notificationType,
      'action_url': actionUrl,
      'is_read': isRead,
      'is_pushed': isPushed,
      'created_at': createdAt.toIso8601String(),
      'expires_at': expiresAt?.toIso8601String(),
      'read_at': readAt?.toIso8601String(),
    };
  }
}

class NotificationResponse {
  final List<AppNotification> notifications;
  final int unreadCount;

  NotificationResponse({
    required this.notifications,
    required this.unreadCount,
  });

  factory NotificationResponse.fromJson(Map<String, dynamic> json) {
    return NotificationResponse(
      notifications: (json['notifications'] as List<dynamic>?)
              ?.map((n) => AppNotification.fromJson(n))
              .toList() ??
          [],
      unreadCount: json['unread_count'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'notifications': notifications.map((n) => n.toJson()).toList(),
      'unread_count': unreadCount,
    };
  }
}

File: backend/internal/domain/models/leaderboard.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// LeaderboardEntry represents a single entry in a leaderboard
type LeaderboardEntry struct {
	Rank              int        `json:"rank"`
	UserID            uuid.UUID  `json:"user_id"`
	Username          string     `json:"username"`
	DisplayName       *string    `json:"display_name,omitempty"`
	AvatarURL         *string    `json:"avatar_url,omitempty"`
	Points            int64      `json:"points"`
	MasteryPercentage *float64   `json:"mastery_percentage,omitempty"`
}

// LeaderboardResponse represents a leaderboard with metadata
type LeaderboardResponse struct {
	Scope      string              `json:"scope"` // daily, weekly, monthly, all_time
	CategoryID *uuid.UUID          `json:"category_id,omitempty"`
	Entries    []LeaderboardEntry  `json:"entries"`
	UserRank   *int                `json:"user_rank,omitempty"`
	TotalUsers int                 `json:"total_users"`
}

// LeaderboardSnapshot represents a saved leaderboard state
type LeaderboardSnapshot struct {
	ID           uuid.UUID  `json:"id" db:"id"`
	UserID       uuid.UUID  `json:"user_id" db:"user_id"`
	CategoryID   *uuid.UUID `json:"category_id,omitempty" db:"category_id"`
	Points       int64      `json:"points" db:"points"`
	Rank         int        `json:"rank" db:"rank"`
	SnapshotType string     `json:"snapshot_type" db:"snapshot_type"` // daily, weekly, monthly, all_time
	SnapshotDate time.Time  `json:"snapshot_date" db:"snapshot_date"`
	CreatedAt    time.Time  `json:"created_at" db:"created_at"`
}

// Notification represents a push notification
type Notification struct {
	ID               uuid.UUID  `json:"id" db:"id"`
	UserID           uuid.UUID  `json:"user_id" db:"user_id"`
	Title            string     `json:"title" db:"title"`
	Body             string     `json:"body" db:"body"`
	NotificationType string     `json:"notification_type" db:"notification_type"`
	ActionURL        *string    `json:"action_url,omitempty" db:"action_url"`
	IsRead           bool       `json:"is_read" db:"is_read"`
	IsPushed         bool       `json:"is_pushed" db:"is_pushed"`
	CreatedAt        time.Time  `json:"created_at" db:"created_at"`
	ExpiresAt        *time.Time `json:"expires_at,omitempty" db:"expires_at"`
	ReadAt           *time.Time `json:"read_at,omitempty" db:"read_at"`
}

// NotificationResponse represents the list of notifications
type NotificationResponse struct {
	Notifications []Notification `json:"notifications"`
	UnreadCount   int            `json:"unread_count"`
}

// RegisterDeviceRequest is the payload for registering FCM device token
type RegisterDeviceRequest struct {
	FCMToken   string `json:"fcm_token" validate:"required"`
	DeviceType string `json:"device_type" validate:"required,oneof=ios android"`
}

// UserDeviceToken represents a user's device for push notifications
type UserDeviceToken struct {
	ID         uuid.UUID  `json:"id" db:"id"`
	UserID     uuid.UUID  `json:"user_id" db:"user_id"`
	FCMToken   string     `json:"fcm_token" db:"fcm_token"`
	DeviceType string     `json:"device_type" db:"device_type"`
	DeviceID   *string    `json:"device_id,omitempty" db:"device_id"`
	IsActive   bool       `json:"is_active" db:"is_active"`
	CreatedAt  time.Time  `json:"created_at" db:"created_at"`
	LastUsed   time.Time  `json:"last_used" db:"last_used"`
}

// UserStreak represents a user's activity streak
type UserStreak struct {
	ID               uuid.UUID  `json:"id" db:"id"`
	UserID           uuid.UUID  `json:"user_id" db:"user_id"`
	CategoryID       *uuid.UUID `json:"category_id,omitempty" db:"category_id"` // NULL for global
	CurrentStreak    int        `json:"current_streak" db:"current_streak"`
	LongestStreak    int        `json:"longest_streak" db:"longest_streak"`
	LastActivityDate *time.Time `json:"last_activity_date,omitempty" db:"last_activity_date"`
	UpdatedAt        time.Time  `json:"updated_at" db:"updated_at"`
}
